catalog/dags/data_refresh/remap_table_indices.py

```diff
@@ -0,0 +1,137 @@
+    index_name: str
+    temp_index_name: str
+    index_def: str
+
+
+def _transform_index_def(
+    existing_index_def: str, temp_table_name: str, table_name: str
+) -> str:
+    """
+    Given a CREATE INDEX statement for an index on the existing media table, return a
+    transformed statement that can be run to apply an identical index to the temp table.
+    """
+    is_pk = "(id)" in existing_index_def
+    tokens = existing_index_def.split(" ")
+
+    # Avoid a naming collision with the existing index by prepending
+    # `temp_import` to the index name. Later when the temp table is promoted,
+    # and the current media table is deleted, the indices will be restored
+    # to their original names.
+    # The index name always follows the word 'INDEX' in the def,
+    # e.g. `CREATE [UNIQUE] INDEX {index_name}`
+    index_name_idx = tokens.index("INDEX") + 1
+    old_index_name = tokens[index_name_idx]
+    # For most indices, we just prefix the index name with `temp_import_`. The pk
+    # index is automatically created at an earlier step and has a different
+    # naming convention.
+    temp_index_name = (
+        f"temp_import_{old_index_name}"
+        if not is_pk
+        else f"temp_import_{table_name}_pkey"
+    )
+    tokens[index_name_idx] = temp_index_name
+
+    # Update the table name to the temp table. The table name always follows
+    # the word 'ON' in the def, e.g. `ON public.audio`
+    table_name_idx = tokens.index("ON") + 1
+    schema_name, table_name = tokens[table_name_idx].split(".")
+    tokens[table_name_idx] = f"{schema_name}.{temp_table_name}"
+
+    return TableIndex(old_index_name, temp_index_name, " ".join(tokens))
+
+
+@task(map_index_template="{{ task.op_kwargs['table_name'] }}")
+def transform_index_defs(
+    existing_index_defs: list[str], temp_table_name: str, table_name: str
+) -> list[TableIndex]:
+    """
+    Given a CREATE INDEX statement for an index on the existing media table, return a
+    transformed statement that can be run to apply an identical index to the temp table.
+    """
+    return [
+        _transform_index_def(index_def, temp_table_name, table_name)
+        for index_def in existing_index_defs
+    ]
+
+
+@task(map_index_template="{{ task.op_kwargs['table_name'] }}")
+def create_table_indices(
+    postgres_conn_id: str, table_name: str, index_configs: list[TableIndex]
+):
+    logger.info(f"Creating indices for `{table_name}`.'")
+    for index_config in index_configs:
+        if "(id)" in index_config.index_def:
+            # Skip the primary key index, as this already exists
+            logger.info(f"Skipping adding {index_config.index_name} index.")
+            continue
+
+        # Create the index
+        run_sql.function(
+            postgres_conn_id=postgres_conn_id, sql_template=index_config.index_def
+        )
+
+    return index_configs
+
+
+def _is_foreign_key(_statement, table):
+    return f"REFERENCES {table}(" in _statement
```

Comment:
Nit: would it be possible to move this function above the `@task` and `@task_group` definitions? IDEs make these functions much easier to find but it does seem odd sandwiched here in between operators 😅

Reply:
This should actually be deleted, it's part of `remap_table_constraints` and was accidentally duplicated/left in when I refactored them out into separate files! I'll remove it :)